contracts/oracle/src/contract/feed/mod.rs

```diff
@@ -0,0 +1,373 @@
+use platform::batch::Batch;
+use sdk::{
+    cosmwasm_ext::Response,
+    cosmwasm_std::{Addr, Storage, Timestamp},
+};
+use serde::de::DeserializeOwned;
+use std::marker::PhantomData;
+use swap::{SwapGroup, SwapTarget};
+
+mod price_querier;
+use price_querier::FedPrices;
+
+mod leg_cmd;
+use leg_cmd::LegCmd;
+
+pub struct Feeds<OracleBase> {
+    feeds: PriceFeeds<'static>,
+    _base: PhantomData<OracleBase>,
+}
+
+impl<OracleBase> Feeds<OracleBase>
+where
+    OracleBase: Currency + DeserializeOwned,
+{
+    pub(crate) fn with(config: Config) -> Self {
+        Self {
+            feeds: PriceFeeds::new("market_price", config),
+            _base: PhantomData,
+        }
+    }
+
+    pub(crate) fn feed_prices(
+        &self,
+        storage: &mut dyn Storage,
+        block_time: Timestamp,
+        sender_raw: &Addr,
+        prices: &[SpotPrice],
+    ) -> Result<(), ContractError> {
+        let tree = SupportedPairs::<OracleBase>::load(storage)?;
+        if prices.iter().any(|price| {
+            !tree.swap_pairs_df().any(
+                |SwapLeg {
+                     from,
+                     to: SwapTarget { target: to, .. },
+                 }| {
+                    price.base().ticker() == &from && price.quote().ticker() == &to
+                },
+            )
+        }) {
+            return Err(ContractError::UnsupportedDenomPairs {});
+        }
+
+        self.feeds.feed(storage, block_time, sender_raw, prices)?;
+
+        Ok(())
+    }
+
+    pub(crate) fn calc_prices(
+        &self,
+        storage: &dyn Storage,
+        at: Timestamp,
+        total_feeders: usize,
+        currencies: &[SymbolOwned],
+    ) -> Result<Vec<SpotPrice>, ContractError> {
+        let tree: SupportedPairs<OracleBase> = SupportedPairs::load(storage)?;
+        let mut prices = vec![];
+        for currency in currencies {
+            let price = self.calc_price(&tree, storage, currency, at, total_feeders)?;
+            prices.push(price);
+        }
+        Ok(prices)
+    }
+
+    pub fn all_prices_iter<'a>(
+        self,
+        storage: &'a dyn Storage,
+        swap_pairs_df: impl Iterator<Item = SwapLeg> + 'a,
+        at: Timestamp,
+        total_feeders: usize,
+    ) -> Result<impl Iterator<Item = BasePrice<SwapGroup, OracleBase>> + 'a, ContractError> {
+        let cmd = LegCmd {
+            price_querier: FedPrices {
+                feeds: self.feeds,
+                storage,
+                at,
+                total_feeders,
+            },
+            stack: vec![],
+        };
+        let res = swap_pairs_df
+            .scan(cmd, |cmd, leg| {
+                let res = currency::visit_any_on_tickers::<SwapGroup, SwapGroup, _>(
+                    &leg.from,
+                    &leg.to.target,
+                    cmd,
+                )
+                .expect("price calculation error");
+                Some(res)
+            })
+            // TODO: process errors
```

Comment:
This is part of the alarms delivery call path which is a subject of this PR. Any idea if/how this should be processed?